src/test/ui/lint/issue-70819-dont-override-forbid-in-same-scope.stderr

```diff
@@ -14,16 +14,5 @@ LL | #![forbid(forbidden_lint_groups)]
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #81670 <https://github.com/rust-lang/rust/issues/81670>
 
-error: deny(unused) incompatible with previous forbid
-  --> $DIR/issue-70819-dont-override-forbid-in-same-scope.rs:21:13
-   |
-LL |     #![forbid(unused)]
-   |               ------ `forbid` level set here
-LL |     #![deny(unused)]
-   |             ^^^^^^ overruled by previous forbid
-   |
-   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
-   = note: for more information, see issue #81670 <https://github.com/rust-lang/rust/issues/81670>
```

Comment:
Looking at this case, it seems to me that the difference is that the first lint is showing the `note` for "lint level defined here" while this one isn't, _and_ the `self.span` might not match the diagnostic's "primary span". An option to avoid issues to special case the lints `hash` calculation, and possibly include the first subdiagnostic in the calculation, that way this one wouldn't disappear, and regular errors wouldn't ever deduplicate either (we have more control over those already).

Reply:
Do you mean we add the first sub-diagnostic if possible?

```rust
    /// Fields used for Hash, and PartialEq trait
    fn key(
        &self,
    ) -> (&Level, &Vec<(String, Style)>, &Option<DiagnosticId>, &MultiSpan, &Vec<CodeSuggestion>)
    {
        (&self.level, &self.message, &self.code, &self.span, &self.suggestions, &self.children.first_or_none  //pesucode
    }
}

impl Hash for Diagnostic {
    fn hash<H>(&self, state: &mut H)
    where
        H: Hasher,
    {
        self.key().hash(state);
    }
}
```

I think we can not remove the duplicated later diagnostic in the case https://github.com/rust-lang/rust/issues/88256.
Since the first diagnostic contains one sub-diagnostic and the second one doesn't have a sub-diagnostic, then we can not avoid the second one because of different hash values.

Reply:
whether we should try to merge those sub diagnostic, since their spans are the same, show in different lints only with different notes seems redundant.

Reply:
I might be wrong, but I think that in all cases there will be at least one subdiagnostic, with the info shown on the first "window" of the error, and using it as part of the deduplication should still work for the case you're trying to tackle, because they only differ on the second subdiagnostic (the note).

Reply:
For the case I'm tracking, there second one doesn't have any subdiagnostics, 
Details are here: https://github.com/rust-lang/rust/issues/88256#issuecomment-908368823.

When the second one is added, this line of code calculated 'refresh' false, so it does not have a note: https://github.com/rust-lang/rust/blob/master/compiler/rustc_session/src/session.rs#L507

Reply:
For this case, currently rustc show the error like this:
![image](https://user-images.githubusercontent.com/230646/131592269-fc9c1dac-5e36-405b-8e29-a597ed717a9a.png)

And we change it to this:
![image](https://user-images.githubusercontent.com/230646/131592294-58f6417b-dbfe-4e3c-8718-1b0d6e1ccf21.png)

As we can see, it's more cleaner, the second diagnostic does not provide any more info but only lacks one note. 
Any other reason we must emit it?

Reply:
I think that the least likely course of action here would be to calculate the hash key using _every_ field and subdiagnostic, except for `note`s with the text "the lint level is defined here". That way we are special casing lints while leaving every other error the same. If we do that, we should add a comment to the test files being affected by this pointing at where that deduplication is being performed, so that if a future change modifies the `note` text, the reviewer/contributor will have a chance to figure out how to fix the deduplication. The other (cleaner) alternative is to modify the `Diagnostic` to have a field telling us whether it is from a lint, and if it is, ignore notes.

Reply:
The first solution is not good, since we need to manually record which notes are ignored.
As for the second solution, except from lint, `Diagnostic` may have any other sources？
Seems for lint, `Diagnostics` is build with: 

```rust
let mut err = lint.build(&span_msg);
...
err.emit()
```

Reply:
Yeah, for the second solution you only need to modify how `Diagnostic` is built in `build`.

Reply:
How about this? 
https://github.com/rust-lang/rust/pull/88493/commits/e1899bbff970283899f797cb14c95049eae3f1b9

Reply:
Make it cleaner, https://github.com/rust-lang/rust/pull/88493/commits/9b8741971c995e0666c07a4c20b4564941c32e06

Reply:
@estebank Test cases are fixed.